### PR TITLE
GA: Promote Receiver API to `notification.toolkit.fluxcd.io/v1`

### PR DESCRIPTION
**Pull request by makkes**
## API changes

The `Receiver` kind was promoted from v1beta2 to v1 (GA). All other kinds of the `notification.toolkit.fluxcd.io` group stay at version v1beta2.

The `receivers.notification.toolkit.fluxcd.io` CRD contains the following versions: 
- v1 (storage version)
- v1beta2 (deprecated)
- v1beta1 (deprecated)

### Upgrade

The `Receiver` v1 API is backwards compatible with v1beta2.

To upgrade from v1beta2, after deploying the new CRD and controller, set  `apiVersion: notification.toolkit.fluxcd.io/v1` in the YAML files that contain `Receiver` definitions. Bumping the API version in manifests can be done gradually. It is advised to not delay this procedure as the beta versions will be removed after 6 months.

refs #436

Tasks tbd before merging:

- [x] update `SOURCE_VER` in Makefile to v1.0.0
- [x] squash commits into reasonable groups

### CODE REVIEW

**Comment by stefanprodan** [comment]
@makkes please fix the "the object has been modified; please apply your changes to the latest version and try again", we need to reload the objects in tests before an update.

File: config/crd/bases/notification.toolkit.fluxcd.io_receivers.yaml
Comment (stefanprodan) [inline config/crd/bases/notification.toolkit.fluxcd.io_receivers.yaml:6]: Please delete the controller-gen binary and rerun make to use the latest version

**Comment by makkes** [comment]
Manual upgrade testing:

1. Bootstrap Flux 0.41.2
2. Put a Receiver v1beta2 manifest in Git and wait for it to be reconciled
3. Build the n-c container image from this PR and make the image available to the cluster
4. Put the Receiver CRD manifest from this PR into the bootstrap repo and patch the n-c Deployment to refer to the newly built image
5. Push the changes and wait for the flux-system Kustomization to get ready and the new n-c Pod to be spun up
6. Verify the Receiver object in the cluster is still ready
7. Check the managed fields of the Receiver
8. Update the Receiver manifest in Git to API version v1
9. Verify the Receiver object in the cluster is still ready
10. Check the managed fields of the Receiver

Results:

- After step 6, Receiver is still ready
- Managed fields after CRD and n-c upgrade:
  ```
  managedFields:
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:kustomize.toolkit.fluxcd.io/name: {}
          f:kustomize.toolkit.fluxcd.io/namespace: {}
      f:spec:
        f:events: {}
        f:resources: {}
        f:secretRef:
          f:name: {}
        f:type: {}
    manager: kustomize-controller
    operation: Apply
    time: "2023-03-28T10:30:33Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"finalizers.fluxcd.io": {}
    manager: notification-controller
    operation: Update
    time: "2023-03-28T10:29:40Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions: {}
        f:observedGeneration: {}
        f:url: {}
        f:webhookPath: {}
    manager: notification-controller
    operation: Update
    subresource: status
    time: "2023-03-28T10:30:33Z"
  ```
- Managed fields after step 10:
  ```

  managedFields:
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:kustomize.toolkit.fluxcd.io/name: {}
          f:kustomize.toolkit.fluxcd.io/namespace: {}
      f:spec:
        f:events: {}
        f:resources: {}
        f:secretRef:
          f:name: {}
        f:type: {}
    manager: kustomize-controller
    operation: Apply
    time: "2023-03-28T10:30:33Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"finalizers.fluxcd.io": {}
    manager: notification-controller
    operation: Update
    time: "2023-03-28T10:29:40Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions: {}
        f:observedGeneration: {}
        f:url: {}
        f:webhookPath: {}
    manager: notification-controller
    operation: Update
    subresource: status
    time: "2023-03-28T10:30:33Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:reconcile.fluxcd.io/requestedAt: {}
    manager: flux
    operation: Update
    time: "2023-03-28T10:54:08Z"
  - apiVersion: notification.toolkit.fluxcd.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:lastHandledReconcileAt: {}
    manager: notification-controller
    operation: Update
    subresource: status
    time: "2023-03-28T10:54:08Z"
  ```

File: docs/spec/v1beta2/events.md
Comment (somtochiama) [inline docs/spec/v1beta2/events.md:4]: Shouldn't the `v1beta2` docs all have the same menu weight?

Comment (makkes) [reply docs/spec/v1beta2/events.md:4]: The menu weight here defines the order of the page in the menu at https://deploy-preview-1427--fluxcd.netlify.app/flux/components/notification/ and the order should always be: controller options (maintained in the website repo), alerts, events, receivers, providers, API reference.

File: docs/spec/v1/receivers.md
Comment (makkes) [reply docs/spec/v1/receivers.md:870]: Good catch.

File: api/v1/receiver_types.go
Comment (stefanprodan) [reply api/v1/receiver_types.go:72]: No, this is required to generate the unique receiver endpoint.

**Comment by makkes** [comment]
> We may need to update https://github.com/fluxcd/notification-controller/blob/main/docs/spec/README.md, similar to how it's updated in SC [fluxcd/source-controller@d905985#diff-c0be2b6354cdd4c26a5c58880a73ec4206d21d8e4c67605ddf6c78fb89e56f90](https://github.com/fluxcd/source-controller/commit/d90598583c68ee0bb99112a45d0fc27a32729555#diff-c0be2b6354cdd4c26a5c58880a73ec4206d21d8e4c67605ddf6c78fb89e56f90) .

I'm not sure I understand the motivation for removing the paragraphs from the README in s-c and the commit message doesn't mention any. Do we not want this documentation to be available anywhere? I don't really mind but would like to understand how to adapt the README here in n-c accordingly.

**Comment by stefanprodan** [comment]
@makkes can you please redo the PR description to document the changes and upgrade procedure? This will help with the changelog.

Examples:
- https://github.com/fluxcd/source-controller/pull/1056
- https://github.com/fluxcd/kustomize-controller/pull/822

**Comment by makkes** [comment]
> v1beta2 `CrossNamespaceObjectReference` is used by `EventServer` in [internal/server/event_handlers.go](https://github.com/fluxcd/notification-controller/blob/249f0d1276e98060ff423ff0303d6b8d76648663/internal/server/event_handlers.go#L289). Similar to how in SC all the non-v1 objects refer to the v1 `Artifact`, I think we can do the same here and use v1 `CrossNamespaceObjectReference`.

The `CrossNamespaceObjectReference` used in EventServer is directly consumed from the `Alert` spec, that's why it's a v1beta2 reference, not a v1 one.

**Comment by makkes** [comment]
> Based on the last comment, all the API definitions in `api/v1beta2` that have reference to `CrossNamespaceObjectReference` should import v1 API and refer to it, similar to how it's done for all the `Artifact` references in SC v1beta2 API.

I must admit I find it strange to not confine all v1beta2 types to their respective Go package but rather reach out to definitions in another API version. This puts a lot of cognitive burden on developers that are working with the v1 types because they might be used in older versions of the API.

**Comment by stefanprodan** [comment]
> I must admit I find it strange to not confine all v1beta2 types to their respective Go package but rather reach out to definitions in another API version. This puts a lot of cognitive burden on developers that are working with the v1 types because they might be used in older versions of the API.

It's only package `github.com/notification-controller/api` so the moment you update the package, all the API versions get imported, the same way it works for Kubernetes apimachinery. What we are doing is similar to Kubernetes Deployments back when Pod Spec was at v1 and Deployments were beta, the beta structs where using the v1 types.

File: api/v1/condition_types.go
Comment (makkes) [reply api/v1/condition_types.go:31]: I was afraid to use the condition types across versions but at this point I don't really mind.

File: docs/spec/v1/README.md
Comment (makkes) [reply docs/spec/v1/README.md:11]: Yeah, I don't think this link belongs here tbh. I'll just remove it. It's confusing.

**Comment by makkes** [comment]
Tasks tbd before merging:

- [ ] update `SOURCE_VER` in Makefile to v1.0.0

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @makkes and @darkowlzz 

PS. We need to wait for SC release, update the API refs in here, then merge.

**Pull request merged**
